subsys/net/lib/sntp/sntp.c

```diff
@@ -0,0 +1,266 @@
+	}
+
+	if (epoch_time) {
+		u64_t ts = (u64_t)ntohl(pkt->tx_tm_s);
+
+		/* Check if most significant bit is set */
+		if ((ts >> 31) != 0) {
+			/* UTC time is reckoned from 0h 0m 0s UTC
+			 * on 1 January 1900.
+			 */
+			if (ts >= OFFSET_1970_JAN_1) {
+				*epoch_time = ts - OFFSET_1970_JAN_1;
+			} else {
+				return -EINVAL;
+			}
+		} else {
+			/* UTC time is reckoned from 6h 28m 16s UTC
+			 * on 7 February 2036.
+			 */
+			*epoch_time = ts + 0x100000000 - OFFSET_1970_JAN_1;
+		}
+	}
+
+	return 0;
+}
+
+static void sntp_recv_cb(struct net_app_ctx *ctx, struct net_pkt *pkt,
+			int status, void *user_data)
+{
+	struct sntp_ctx *sntp = NULL;
+
+	if (!pkt) {
+		return;
+	}
+
+	if (status < 0) {
+		return;
+	}
+
+	if (!user_data) {
+		goto error_exit;
+	}
+	sntp = (struct sntp_ctx *)user_data;
+
+	if (net_pkt_appdatalen(pkt) != sizeof(struct sntp_pkt)) {
+		goto error_exit;
+	}
+
+	if (!net_pkt_appdata(pkt)) {
+		goto error_exit;
+	}
+
+	k_fifo_put(&sntp->recv_q, pkt);
+	return;
+
+error_exit:
+	if (ctx) {
+		k_fifo_cancel_wait(&sntp->recv_q);
+	}
+	net_pkt_unref(pkt);
+}
+
+static u32_t get_uptime_in_sec(void)
+{
+	u64_t time;
+
+	k_enable_sys_clock_always_on();
+	time = k_uptime_get_32();
+	k_disable_sys_clock_always_on();
+
+	return time / 1000;
+}
+
+int sntp_init(struct sntp_ctx *ctx, const char *srv_addr, u16_t srv_port,
+		u32_t timeout)
+{
+	int rv;
+
+	if (!ctx) {
+		return -EFAULT;
+	}
+
+	memset(ctx, 0, sizeof(struct sntp_ctx));
+
+	k_fifo_init(&ctx->recv_q);
+
+	rv = net_app_init_udp_client(&ctx->net_app_ctx, NULL, NULL, srv_addr,
+					srv_port, ctx->timeout, ctx);
+	if (rv < 0) {
+		NET_DBG("Failed to init udp client");
+		return rv;
+	}
+
+	rv = net_app_set_cb(&ctx->net_app_ctx, NULL, sntp_recv_cb, NULL, NULL);
+	if (rv < 0) {
+		NET_DBG("Failed to set net app callback");
+		return rv;
+	}
+
+	ctx->timeout = timeout;
+	ctx->is_init = true;
+
+	return 0;
+}
+
+int sntp_request(struct sntp_ctx *ctx, u64_t *epoch_time)
+{
+	struct sntp_pkt tx_pkt = {0};
+	struct sntp_pkt rx_pkt = {0};
+	struct net_pkt *pkt = NULL;
+	int rv = 0;
+	u64_t tmp = 0;
+	u16_t offset = 0;
+
+	if (!ctx) {
+		return -EFAULT;
+	}
+
+	if (!ctx->is_init) {
+		return -EINVAL;
+	}
+
+	/* prepare request pkt */
+	tx_pkt.li = 0;
+	tx_pkt.vn = SNTP_VERSION_NUMBER;
+	tx_pkt.mode = SNTP_MODE_CLIENT;
+	ctx->expected_orig_ts = get_uptime_in_sec() + OFFSET_1970_JAN_1;
+	tx_pkt.tx_tm_s = htonl(ctx->expected_orig_ts);
+
+	rv = net_app_connect(&ctx->net_app_ctx, ctx->timeout);
+	if (rv < 0) {
+		NET_DBG("Failed to connect");
+		return rv;
+	}
+
+	rv = net_app_send_buf(&ctx->net_app_ctx, (u8_t *)&tx_pkt,
+				sizeof(tx_pkt),
+				&SNTP_CTX_SRV_SOCKADDR(ctx),
+				sizeof(SNTP_CTX_SRV_SOCKADDR(ctx)),
+				ctx->timeout, NULL);
+	if (rv < 0) {
+		NET_DBG("Failed to send request");
+		return rv;
+	}
+
+	pkt = k_fifo_get(&ctx->recv_q, ctx->timeout);
+	if (!pkt) {
+		return -EAGAIN;
+	}
+
+	/* copy to buf */
+	offset = net_pkt_get_len(pkt) - net_pkt_appdatalen(pkt);
```

Comment:
Again, code duplicate from bsd sockets. The whole idea of developing bsd sockets API was to save apps from the need to duplicate that code over and over again.